cardano-cli/src/Cardano/CLI/EraBased/Run/StakePool.hs

```diff
@@ -221,19 +227,73 @@ runStakePoolMetadataHashCmd
+            . newExceptT
+            $ readByteStringFile poolMetadataFile
+        StakePoolMetadataURL urlText ->
+          fetchURLToStakePoolCmdError $ getByteStringFromURL allSchemas urlText
 
     (_metadata, metadataHash) <-
       firstExceptT StakePoolCmdMetadataValidationError
         . hoistEither
         $ validateAndHashStakePoolMetadata metadataBytes
-    case mOutFile of
-      Nothing -> liftIO $ BS.putStrLn (serialiseToRawBytesHex metadataHash)
-      Just (File fpath) ->
-        handleIOExceptT (StakePoolCmdWriteFileError . FileIOError fpath) $
-          BS.writeFile fpath (serialiseToRawBytesHex metadataHash)
+
+    case hashGoal of
+      CheckStakePoolMetadataHash expectedHash
+        | metadataHash /= expectedHash ->
+            left $ StakePoolCmdHashMismatchError expectedHash metadataHash
+        | otherwise -> liftIO $ putStrLn "Hashes match!"
+      StakePoolMetadataHashToFile outFile -> writeOutput (Just outFile) metadataHash
+      StakePoolMetadataHashToStdout -> writeOutput Nothing metadataHash
+   where
+    writeOutput :: Maybe (File () Out) -> Hash StakePoolMetadata -> ExceptT StakePoolCmdError IO ()
+    writeOutput mOutFile metadataHash =
+      case mOutFile of
+        Nothing -> liftIO $ BS.putStrLn (serialiseToRawBytesHex metadataHash)
+        Just (File fpath) ->
+          handleIOExceptT (StakePoolCmdWriteFileError . FileIOError fpath) $
+            BS.writeFile fpath (serialiseToRawBytesHex metadataHash)
+
+    fetchURLToStakePoolCmdError
+      :: ExceptT FetchURLError IO BS.ByteString -> ExceptT StakePoolCmdError IO BS.ByteString
+    fetchURLToStakePoolCmdError = withExceptT StakePoolCmdFetchURLError
+
+-- | Check the hash of the anchor data against the hash in the anchor if
+-- checkHash is set to CheckHash.
+carryHashChecks
+  :: PotentiallyCheckedAnchor StakePoolMetadataReference StakePoolMetadataReference
+  -- ^ The information about anchor data and whether to check the hash (see 'PotentiallyCheckedAnchor')
+  -> ExceptT StakePoolCmdError IO ()
+carryHashChecks potentiallyCheckedAnchor =
+  case pcaMustCheck potentiallyCheckedAnchor of
+    CheckHash -> do
+      let url = toUrl $ stakePoolMetadataURL anchor
+      metadataBytes <-
+        withExceptT
+          StakePoolCmdFetchURLError
+          (getByteStringFromURL httpsAndIpfsSchemas url)
+
+      let expectedHash = stakePoolMetadataHash anchor
+
+      (_metadata, metadataHash) <-
+        firstExceptT StakePoolCmdMetadataValidationError
+          . hoistEither
+          $ validateAndHashStakePoolMetadata metadataBytes
+
+      when (metadataHash /= expectedHash) $
+        left $
+          StakePoolCmdHashMismatchError expectedHash metadataHash
+    TrustHash -> pure ()
+ where
+  anchor = pcaAnchor potentiallyCheckedAnchor
+
+  toUrl :: Text -> L.Url
+  toUrl t =
+    let l = BS.length (encodeUtf8 t)
+     in fromMaybe (error "Internal Error: length of URL was miscalculated") $ L.textToUrl l t
```

Comment:
Hum sneaking an `error` here is not good practice :slightly_smiling_face:, please put this function in the `ExceptT StakePoolCmdError` monad, and add an error case to `StakePoolCmdError`.

Reply:
The reason I did it like this is that it is an error that should not happen unless there is a bug in the code, a "this should never happen" kind of thing. But let me see if I can get rid of it altogether

Reply:
Yes, much simpler actually. The alternative would be to export the `Url` constructor through the `cardano-api` but we can do that later.
Addressed here: https://github.com/IntersectMBO/cardano-cli/pull/932/commits/586049cf819f63ccbd860243d6ca8b442d794deb